### ISSUE TITLE
Remove obsolete/empty databases from mongo db

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1431]
The following databases are either empty or have only empty collections in them:

mwf_dbname
modularforms_2010
quadratic_twists
modforms
WebNewForms

Several others are never referenced by any of the code in LMFDB/lmfdb and have no inventory information in LMFDB/inventory (I will post a list later).

These should definitely be removed from the cloud server (so that they do not appear on http://www.lmfdb.org/api/, for example).  Presumably the empty ones (and possibly others) can/should also be removed from the mongo db on Atkin.

**Comment by JohnCremona** [comment]
I can think of no reason for one of us not to remove these 5 right now.  If they are removed from Warwick what happens with replication?  I don't know how to connect or the cloud replicas.

**Comment by AndrewVSutherland** [comment]
I can take care of the cloud.  If you delete it on Warwick I believe 
this should get propagated automatically to the replicas (but 
@edgarcosta can confirm).

**Comment by AndrewVSutherland** [comment]
@JohnCremona, @edgarcosta The five empty databases listed above have been removed from the cloud mongo db (running on ms.lmfdb.xyz) and no longer appear on www.lmfdb.org/api/

**Comment by JohnCremona** [comment]
You beat me to it, I have removed the first one....

**Comment by AndrewVSutherland** [comment]
@JohnCremona, @davidfarmer  Do we need the databases knowledge_5, knowledge_6, knowledge_7, knowledge_8, knowledge_9, and knowledge_tmp?  It looks to me like the knowledge database contains everything that is in these.

**Comment by AndrewVSutherland** [comment]
@sehlen Do we still need the databases modularforms and modularforms_raw?  It looks like the only place where modularforms is referenced is in test_root.py (which should presumably be changed to modularforms2).

**Comment by edgarcosta** [comment]
@AndrewVSutherland The knowledge_* stuff is my fault. Cleaning it now.

**Comment by AndrewVSutherland** [comment]
@JohnCremona I presume it makes sense to remove the "limbo" database from the cloud server (and possibly Warwick as well?)

**Comment by jwj61** [comment]
Yes.  It was the original Artin representation database, but has been superseded by artin.  So yes, it can be deleted.

**Comment by JohnCremona** [comment]
OK then I'll do the honours at the Warwick end,

**Comment by edgarcosta** [comment]
What about "MaassWaveForm" (without the "s")?

There are 2 files that mention it:
* never ends up using it : https://github.com/LMFDB/lmfdb/blob/master/lmfdb/modular_forms/maass_forms/maass_waveforms/backend/mwf_utils.py 
* Wants to check for "MaassWaveForms" instead?: https://github.com/LMFDB/lmfdb/blob/master/lmfdb/test_root.py

**Comment by edgarcosta** [comment]
limbo also dropped at ms.lmfdb.xyz

**Comment by JohnCremona** [comment]
mongo makes it much too easy to create a new database by mistake after a typo.  At least, that was true before we added authentication.

**Comment by fredstro** [comment]
The collection
‘MaassWaveForm’
can be dropped, similarly the collections
‘modularforms_raw’    (this is not really used and was something I was experimenting with) 
and 
‘modularforms’   (this was the predecessor to ‘modularforms2’ and hasn’t been used since 2012)
can both be dropped.

Fredrik

**Comment by edgarcosta** [comment]
Catching!

Done at warwick 
```
lmfdb0:PRIMARY> use MaassWaveForm
switched to db MaassWaveForm
lmfdb0:PRIMARY> db.dropDatabase()
{ "dropped" : "MaassWaveForm", "ok" : 1 }
lmfdb0:PRIMARY> use modularforms_raw
switched to db modularforms_raw
lmfdb0:PRIMARY> db.dropDatabase()
{ "dropped" : "modularforms_raw", "ok" : 1 }
lmfdb0:PRIMARY> use modularforms
switched to db modularforms
lmfdb0:PRIMARY> db.dropDatabase()
{ "dropped" : "modularforms", "ok" : 1 }
```

and ms:
```
> use limbo
switched to db limbo
> db.dropDatabase()
{ "dropped" : "limbo", "ok" : 1 }
> use MaassWaveForm
switched to db MaassWaveForm
> db.dropDatabase()
{ "dropped" : "MaassWaveForm", "ok" : 1 }
> use modularforms_raw
switched to db modularforms_raw
> db.dropDatabase()
{ "dropped" : "modularforms_raw", "ok" : 1 }
> use modularforms
switched to db modularforms
> db.dropDatabase()
{ "dropped" : "modularforms", "ok" : 1 }
````

**Comment by edgarcosta** [comment]
I also took a snapshot before going on this dropDatabase spree...

**Comment by JohnCremona** [comment]
@edgarcosta Good idea.  I was thinking that we also have the weekly dumps, but they are not kept for that long (40 days according to the backup script).  Though I do have some other copies.

**Comment by edgarcosta** [comment]
The db just got much slimmer:
* 1TB using MMAPv1 (e.g. warwick and ms)
* 620GB using using wiredTiger+zlib

**Comment by edgarcosta** [comment]
@JohnCremona we should discuss for how long, and how frequent should we keep backups in the cloud

**Comment by JohnCremona** [comment]
@edgarcosta Your previous comment reminded me of a question you asked me (no idea which thread or issue or what) which I did not understand.  If it's about filesystems used on the Warwick server please include Bober and Schilly (perhaps you did).

**Comment by AndrewVSutherland** [comment]
Does anyone know anything about ap_statistics?  It is not referenced by any code in LMFDB/lmfdb.

**Comment by fredstro** [comment]
It contains data about a(p)’s for newforms of  weight 2, trivial character and where the degree of the coefficient field is 2.
It is data which I and David F. was/are planning to use to analyse a variant of Maeda’s conjecture…   
(The collection ‘madea’ in ‘modularforms2’ contains a subset of collection ap_data on the database ‘ap_statistics) 

I just mongodumped the database so you can safely delete it if you like. 

At some point in the future we would probably like to display statistics for modular forms but we have to think more about exactly what should be (pre-)computed and what should be stored where. 


The reason why there are so many collections in modularforms2 is that the mongo server in warwick (previously Washington I think) was basically the
only database we all had had access to so it was used for a lot of testing, experimenting development and debugging… 

Basically every collection on modularforms2 which is not referenced to somewhere in the lmfdb can probably be safely deleted (@sehlen?). 
 
.

**Comment by AndrewVSutherland** [comment]
@fredstro, @sehlen, So when I updated the modularforms2 data on www.lmfdb.org I copied over the following 11 collections:

dimension_table
dimension_table.chunks
dimension_table.files

webmodformspace
webmodformspace.chunks
webmodformspace.files

webnewforms
webnewforms.chunks
webnewforms.files

webeigenvalues.chunks
webeigenvalues.files

I see  the code also references "webmodformspace_dimension" in emf_utils.py

Are there any other collections that are needed?

**Comment by AndrewVSutherland** [comment]
@fredstro OK, it sounds like I can definitely remove ap_statistics from the cloud, I'll let @JohnCremona decide what to do on the Warwick machines.

**Comment by AndrewVSutherland** [comment]
@davidfarmer is it safe to remove the database Lfunction from the cloud (and possibly Warwick also)?  It looks like the code uses the Lfunctions database (both for new and old formats)

**Comment by JohnCremona** [comment]
As luck would have it one of the tests in test_root checks for MaassWaveForm, so now fails.  Rather than fix this one, we should adapt the test_db function there to check for a complete list of the databases which we actually need.

**Comment by edgarcosta** [comment]
I saw that, and I was going to edit that, but then I noticed that in the master branch it is an empty file:
https://github.com/LMFDB/lmfdb/blob/master/test_root.py

**Comment by JohnCremona** [comment]
Wrong file:  lmfdb/test_root.py is the one.  That empty file can probably be deleted!

**Comment by JohnCremona** [comment]
@jwj61 the test of the link to http://hobbes.la.asu.edu/lmfdb-14/  now fails.  Is that just temporary?  There's a test for it in test_acknowledgements.py

**Comment by jwj61** [comment]
Temporary.  Due to work in the building housing hobbes, all computers have been shut down for a week (and maybe more, but hopefully just a week).

**Comment by edgarcosta** [comment]
> As luck would have it one of the tests in test_root checks for MaassWaveForm, so now fails. Rather than fix this one, we should adapt the test_db function there to check for a complete list of the databases which we actually need.

new issue? 
Perhaps we should first take care of:
https://github.com/LMFDB/lmfdb-inventory
maybe assign the issues to the owners?

**Comment by AndrewVSutherland** [comment]
@edgarcosta fixing test_root.py is something we should probably take care of.  Getting lmfdb-inventory up to date for each database is something we can assign to the owners, but first we need to decide which databases should really be there and actually give them owners.  Currently lmfdb-inventory has a .md file for every database that was in the Warwick mongo db at the time it was created (several of which we just deleted), many of which are just stubs with no owner.  Part of this issue should be fixing this.

**Comment by edgarcosta** [comment]
@AndrewVSutherland I totally agree with you.
However, I don't know a nice and simple way to figure out which DBs we really want to have there.

**Comment by AndrewVSutherland** [comment]
@edgarcosta That is one of the goals of this issue; as noted above one way to check is to search the code for references to the database names.   Here is a list of databases that I know should be on the production cloud mongo db servring data to www.lmfdb.org (ms.lmfdb.xyz)

sato_tate_groups
localfields
elliptic_curves
numberfields
knowledge
siegel_modular_forms
Lattices
hmfs
Lfunctions
modularforms2
MaasWaveForms
transitivegroups
artin

Here is a list of databases that are used in the code and potentially accessible from www.lmfdb.org.  It would be good to confirm their exact status
HTPicard
siegel_modular_forms_experimental
SL2Zsubgroups

Here is a list of additional databases that I know need to be on the Warwick dev/beta database (lmfdb.warwick.ac.uk) but are not needed in the cloud (yet) because they are only used in beta mode (in some cases not in the sidebar yet, but there is code in lmfdb/lmfdb that uses them)

mod_l_eigenenvalues
halfintegralmf
hgm
curve_automorphisms

Here is a list of database that look useful but are not referenced in code.  They may have future use and we might want to keep them on the Warwick machine.  Definitely not needed in the cloud:

abvar (only on Warwick, not in cloud)
ap_statistics (see https://github.com/LMFDB/lmfdb/issues/1431#issuecomment-220794717, not in cloud) 
bmfs (Bianchi modular forms data?  in cloud now, could/should be removed)

**Comment by AndrewVSutherland** [comment]
Just to follow up on this, according to "https://github.com/LMFDB/lmfdb-inventory/blob/master/db-Lfunction.md" we should be able to delete the Lfunction database (everything is in Lfunctions).

Any objections?

**Comment by JohnCremona** [comment]
Not from me.  Two of the collectios are *Test anyway, so if we are sure about the 438 items in LemurellMaassHighDegree then fine.

**Comment by AndrewVSutherland** [comment]
@JohnCremona I am not sure, and in fact I notice that this collection *is* referenced in the code, see https://github.com/LMFDB/lmfdb/search?utf8=%E2%9C%93&q=LemurellMaassHighDegree.

Maybe it would be a good idea to copy this collection into the Lfunctions database?  Or at least ask @davidfarmer about it?

**Comment by AndrewVSutherland** [comment]
And in fact the two "test" collections are also referenced in https://github.com/LMFDB/lmfdb/blob/master/lmfdb/lfunctions/LfunctionDatabase.py

**Comment by JohnCremona** [comment]
OK let's leave them.  L-functions are in a transitional stage so we cannot expect the set of collections to be 100% tidy right now.

**Comment by edgarcosta** [comment]
Should I merge #1373 into this issue?
Most of the databases have those two collections that I believe to be relics of mongodb 2.4.

**Comment by jenpaulhus** [comment]
From Drew's list a couple days ago:

curve_automorphisms is "mine". I'm still a bit fuzzy about the cloud vs beta/development, but pages like
http://www.lmfdb.org/HigherGenus/C/aut/3.96-64.0.2-3-8
use that database.  There is no link on the sidebar yet, even in beta (I want to make sure a couple of other features are in place before then.)

Making a page for it on lmfdb-inventory is on my to do list.

**Comment by AndrewVSutherland** [comment]
@jenpaulhus Thanks for the confirmation.  Given that the pages are accessible (but hidden) on www.lmfdb.org, I think it makes sense to leave these on the "cloud server", by which I mean the mongo db hosted by ms.lmfdb.xyz, which serves data to the web servers hosting www.lmfdb.org (and running the LMFDB application code), all of which are instances in Google's Compute Engine (aka "the cloud"); this is separate from the mongo db hosted by  lmfdb.warwick.ac.uk, which serves data to beta.lmfdb.org  as well as the hosts atkin and lehner at Warwick which are used for development.

**Comment by AndrewVSutherland** [comment]
@edgarcosta Yes I think it makes sense to merge #1373 into this one.  So far we have been focusing on the database level, but the next step is to drill down to collections.

**Comment by lemurell** [comment]
It is safe to delete the Lfunction database. Everything of value is now in Lfunctions. The code that references the collections is obsolete

/Stefan

**Comment by JohnCremona** [comment]
Thanks Stefan.   It would then be helpful to remove the obsolete code, could you do that?   Nothing g is ever really lost it git anyway.

**Comment by AndrewVSutherland** [comment]
With help from @sehlen and @fredstro I was able to remove what I believe are all unnecessary collections from the modularforms2 database in the cloud; only the following collections remain:

dimension_table
webmodformspace webmodformspace.chunks webmodformspace.files
webnewforms webnewforms.chunks webnewforms.files
webeigenvalues.chunks webeigenvalues.files
webchar webchar.chunks webchar.files

Mongo DB now shows 540GB of free space available in this database.  Only about 85GB is currently used.  This will double when we copy over new files (and want to save backups of the old ones temporarily), and it will increase again once we add in the missing data, but my guess is that it will stabilize well below the 622GB currently allocated.  @edgarcosta should we think about dumping and restoring this database at some point to recover the unused space?

I am just about ready to close this issue, the only outstanding things in my mind are (1) the obsolete Lfunction database that is still referenced in the code (see https://github.com/LMFDB/lmfdb/issues/1456), and (2) the fact that there are still a lot of unnecessary collections in the modularforms2 database in Warwick taking up a lot of space (my guess is that about half the space in the mongo db on lmfdb.warwick.ac.uk is taken up by collections in modularforms2 that do not need to be there).  Getting rid of it would speed up backup/restore operations (among other things).

**Comment by edgarcosta** [comment]
@AndrewVSutherland I don't follow what you mean by this:
> Mongo DB now shows 540GB of free space available in this database. 
The disk? and what server are you talking about?

Regarding
> should we think about dumping and restoring this database at some point to recover the unused space?
I believe that Sarunas is done testing the different storage engines, see the *png files in:
https://github.com/edgarcosta/lmfdb-gce/tree/master/ab/ms-bench_n1hcpu4_storage
Thus we are close to just push a new snapshot of the DB to ms.lmfdb.xyz.

**Comment by AndrewVSutherland** [comment]
In mongo on ms.lmfdb.xyz do db.stats and look at the free list size.

On May 29, 2016 6:07:35 PM EDT, edgarcosta <notifications@github.com> wrote:
>@AndrewVSutherland I don't follow what you mean by this:
>> Mongo DB now shows 540GB of free space available in this database. 
>The disk? and what server are you talking about?
>
>Regarding
>> should we think about dumping and restoring this database at some
>point to recover the unused space?
>I believe that Sarunas is done testing the different storage engines,
>see the *png files in:
>https://github.com/edgarcosta/lmfdb-gce/tree/master/ab/ms-bench_n1hcpu4_storage
>Thus we are close to just push a new snapshot of the DB to
>ms.lmfdb.xyz.
>
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/LMFDB/lmfdb/issues/1431#issuecomment-222385599

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by edgarcosta** [comment]
I see.
When we switch the engine in non-human effort way, mongo will take care of all that. We will do this soon.

**Comment by AndrewVSutherland** [comment]
@edgarcosta Sounds good.  I just copied over a new set of modular forms data so that there is currently two copies of evey collection in modularforms2 and there is still 460GB free.

**Comment by edgarcosta** [comment]
@AndrewVSutherland will you also cleanup modularforms2 at warwick?

**Comment by AndrewVSutherland** [comment]
@edgarcosta I think this should be done by @sehlen and @fredstro, there may be collections there that they are using to compute (or recompute) modular forms data and I don't want to risk removing something that they need.  But I know it would free up a lot of space (both in Warwick and on the cloud replicas), so I would certainly encourage them to do so!

**Comment by sehlen** [comment]
@AndrewVSutherland @edgarcosta

For warwick, different rules have to apply as some more collections are used.
After I had a closer look at all of them and consulted with @fredstro, I can delete what is not needed anymore next week or so.
We can also migrate the collections we use for computing modular forms somewhere else;
possibly another mongo database that’s not replicated if you don’t want it in the cloud
(or can you selectively replicate collections?)...

**Comment by edgarcosta** [comment]
There is no need to move them.
However, would be very helpful to have each collection documented here:
https://github.com/LMFDB/lmfdb-inventory/blob/master/db-modularforms2.md
and at the top be clear which collections are needed for the website.

**Comment by sehlen** [comment]
> On Jun 1, 2016, at 13:30, edgarcosta <notifications@github.com> wrote:
> 
> There is no need to move them.
> However, would be very helpful to have each collection documented here:
> https://github.com/LMFDB/lmfdb-inventory/blob/master/db-modularforms2.md <https://github.com/LMFDB/lmfdb-inventory/blob/master/db-modularforms2.md>
> and at the top be clear which collections are needed for the website.
> 
> 
I started this a few days ago and will continue soon.

**Comment by JohnCremona** [comment]
I have no idea how far this has got, and rereading all the above thread it is is extremely confusing to say the least.  Anyone who has not yet bothered to document the collections they use after months of being asked deserves to have them deleted, and anyone who cannot recreate their collections from data held externally to mongo similarly.

**Comment by AndrewVSutherland** [comment]
@JohnCremona while I agree with your sentiment, as it stands, if we unilaterally removed all undocumented collections it would render a substantial portion of www.lmfdb.org non-functional (none of the modular forms collections currently serving data to www.lmfdb.org, including classical MFs, Hilbert MFs, Maass forms, and Siegel modular forms, have been documented).

There are currently 9 open issues on https://github.com/LMFDB/lmfdb-inventory/issues (this also includes some that are currently only relevant to beta.lmfdb.org).  Based on my current understanding, the likely candidates for addressing the modular forms collections are:

@jvoight (HMFs), @sehlen and @fredstro (classical MFs), @fredstro (Maass forms), and @nilsskoruppa (Siegel modular forms).

But please correct me if I am wrong, or if there are other people who should be contacted.

For my own part, in order to help organize/summarize our current situation, I will put together a list of all the databases that currently exist in the cloud or Warwick mongo dbs and what I think their current status is.  I will try to post this later today.

**Comment by JohnCremona** [comment]
@AndrewVSutherland you're a hero.

**Comment by AndrewVSutherland** [comment]
Below is a table summarzing of all the databases the currently exist on the warwick mongo db (excluding local, which is only used by mongo db itself), which you will want to view on github.

This is a first draft that I will update as I get more information.  Any comments on the question-marks and/or proposed deletions would be welcome!

|database|size(GB) prod/beta|status|doc status|code|action|
|---|---|---|---|---|---|
|HTPicard|0.0/0.1|hidden|none|yes|delete?|
|Lattices|0.0/0.5|prod|complete|yes||
|Lfunction|0.0/0.1|delete|none|lyes|delete|
|MaassWaveForms|0.7/2.0|prod|none|yes|document|
|SL2Zsubgroups|0.0/0.1|hidden|none|yes|#1407|
|abvar|0.95/--|???|none|no|delete|
|ap_statistics|0.5/--|???|none|no|delete|
|artin|0.0/1.0|prod|need index info|yes|document|
|bmfs|0.2/0.0|future|none|no|???|
|contrib|1.0/--|beta|none|yes|document|
|curve_automorphisms|0.1/0.0|beta|none|yes|document|
|elliptic_curves|25.9/2.1|prod|complete|yes||
|genus2_curves|51.9/0.1|prod|complete|yes||
|halfintegralmf|0.1/0.0|dev|partial|yes|document|
|hgm|0.1/0.0|beta|none|yes|document|
|hmfs|175.9/20.3|prod|none|yes|document|
|knowledge|0.1/0.0|prod|none|yes|document|
|localfields|0.1/0.0|prod|complete|yes||
|mod_l_eigenvalues|0.1/0.0|dev|none|yes||
|mod_l_galois|0.1/--|dev|none|yes||
|modularforms2|517.7/69.1|prod|none|yes|document|
|numberfields|61.9/3.5|prod|complete|yes||
|sato_tate_groups|0.00/0.08|prod|complete|yes||
|siegel_modular_forms|0.2/0.0|prod|none|yes|document|
|siegel_modular_forms_experimental|2.0/0.2|???|none|yes|???|
|test|0.1/0.0|???|none|no|delete?|
|transitivegroups|0.1/0.0|prod|partial|yes|document|
|upload|1.0/0.1|beta|none|yes|document|
|userdb|0.0/0.0|beta|none|yes|document|

**Comment by JohnCremona** [comment]
I added documentation for bmfs to the inventory.  There *is* code but it is in my own branch, since for 2 years I have not had time to work on this, and did not want to even put it onto beta until I had added some newforms and not just dimension tables.  One reason for that being stalled is that I wanted to have the same layout and code work for all modular forms over number fields, i.e. merge Bianchi with Hilbert, and this has never reached the top of anyone-s to-do list.

**Comment by JohnCremona** [comment]
Can we put a version of this table somewhere where it would be easier to edit?  Perhaps as a file of its own in the inventory repository, linked to from here?
We also need something at the collection level.
In database elliptic_curves:  the curves collection is now redundant since the code (master, beta, prod) uses collection curves2 if is is there, which is is.  The collections curves2.rand and nfcurves.rand were created for testing the new random object functionality.   So I think curves2.rand can replace curves2, but I do not know the correct procedure to do this since the random object function requires the collection name with ".rand" suffix but the elliptic curve code does not want the suffix.

**Comment by AndrewVSutherland** [comment]
@JohnCremona Thanks for the clarification.  I actually think it is better to keep unfinished/untested code out of the lmfdb repository, so I agree with this approach.  My impressions is that there is a fair amount of code currently in LMFDB/lmfdb that is either obsolete or never has not actually been run in several years (and would be very unlikely to work if it were because a lot has changed since then).

Unused and/or obsolete code sitting in the repository the lmfdb repository still imposes a maintenance burden -- every time we identify some issue that potentially affects the entire code base (e.g. switching to a new version of sage or python, removing deprecated usages like "import *" or eval, identifying every place a given function that we want to change is used) we have to look at and think about this code.

I mention this because it is relevant to some of the databases above, e.g. HTPicard -- there is code related to this database in lmfdb/modular_forms/maass_forms/picard, but my impression is that most of it has not been updated in 3-4 years.  If we are going to remove the HTPicard database we should remove this code.  If we are not, we should make sure that it actually runs with the current version of the lmfdb.

**Comment by sanni85** [comment]
I will document mod_l_eigenvalues and mod_l_galois as soon as possible.

I think that the database abvar should not be deleted: it is the database created at AIM by @christellevincent @kedlaya @roed314

**Comment by AndrewVSutherland** [comment]
@JohnCremona Regarding https://github.com/LMFDB/lmfdb/issues/1431#issuecomment-225532695, yes I will put it in LMFDB/inventory and I do want to add collection level information, I just have not had time to do so.  BTW, when I listed documentation as "complete" I did not take the time to verify that every collection is document, I still want to do this.

The *.rand collections are not meant to replace *, they are auxiliary collections (think of them as indexes) that are automatically by the create_random_object_index function and contain only two fields, "_id" and "num".  They are used by "random_object_from_collection" to speed up random object access.  I will add documentation to the inventory noting this.

The collection *.rand should be present if and only if * is present and is never meant to replace *.

**Comment by AndrewVSutherland** [comment]
@JohnCremona To replace curves with curves2, just rename curves2 to curves and rename curves2.rand to curves.rand.  The existence or (non-existence) of curves.rand has no impact on any of the elliptic curves code, only the functions random_object_from_collection and random_value_from_collection care about it.  The only impact that not having it would be that random object access would slow down.

**Comment by JohnCremona** [comment]
OK, I understand now.  I wrongly thought that (e.g.) curves2.rand contained everything in curves2 plus one additional field.  Sorry!  This is fine then, I was afraid that we were using double the space (but of course not).
I am still very uneasy about making any changes like this since I do not understand how replication will be affected.  Should I just ignore that and assume that it will do the right thing?

**Comment by AndrewVSutherland** [comment]
@JohnCremona  Yes!  It replicates the operations (so a rename gets replicated as a rename).

BTW I just added information about the existing *.rand collections to the index information of all the relevant collections (including nfcurves, which I notice is still missing index info).

**Comment by davidfarmer** [comment]
The maass_forms/picard code, and the associated database, is something
we want in the LMFDB, but there has been no maintenance of it for
years.  It will be easier to resurrect those objects from the existing
code and database than to re-create them from scratch, so I think it
would be better to leave them in.

BTW: those objects are Maass forms on SL(2,\Z[i]).

> I mention this because it is relevant to some of the databases above, e.g. HTPicard
> -- there is code related to this database in lmfdb/modular_forms/maass_forms/picard,
> but my impression is that most of it has not been updated in 3-4 years. If we are
> going to remove the HTPicard database we should remove this code. If we are not, we
> should make sure that it actually runs with the current version of the lmfdb.

**Comment by christellevincent** [comment]
Yes, please do not delete abvar. As Samuele said, this is our new database which we worked on at the workshop in May. After we left we all got busy, but I hope we can push to beta pretty soon if @roed314 has some time, and we'll document the database.

**Comment by JohnCremona** [comment]
@AndrewVSutherland I'll add index info for nfcurves after rebuilding the entire collection with extra data fields, which will also mean updating the collection information.

**Comment by AndrewVSutherland** [comment]
OK, is HTPicard functionality somethings that should be scheduled into 
Release 2.0?  And who should be listed as the contact/maintainer?

**Comment by AndrewVSutherland** [comment]
Thanks @christellevincent and @sanni85!  Who should be listed as the 
contact/maintainer for the abvar database?

**Comment by JohnCremona** [comment]
OK I just did
 
 sage: curves2.rename("curves", dropTarget=True)
 sage: curves2.rand.rename("curves.rand", dropTarget=True)                                            

and once the changes have propagated to the cloud db I can clean up the code to no longer mention curves2.

**Comment by davidfarmer** [comment]
I'd guess HTPicard is beyond 2.0 because I don't think we have
anyone who would be suitable to take care of those.  (Stefan
Leurell could do it, but he already has other things on his plate.)

**Comment by AndrewVSutherland** [comment]
I will update the mongo db on ms.lmfdb.xyz to reflect this change today (and let you know when I do).

**Comment by christellevincent** [comment]
I think David Roe @roed314 should be the contact for abvar. He's the one who knows the backend.

**Comment by AndrewVSutherland** [comment]
@davidfarmer OK, should I list you or Stefan as the contact/maintainer?  I just want someone's name listed as a contact so that the next time someone wonders why this database is there or if there is a maintenance issue that effects this database (e.g. a mongo db upgrade) they know who to contact.

**Comment by JohnCremona** [comment]
The reason why elliptic curves /Q do not work right now on beta is just because of a stupid typo in the code which was written to ensure that the code should work with any combination of curves/curves2 existing.  I'm about to make a pull request which will fix that, can be merged to beta right away, and to prod once the renaming of curves2 -> curve change has propagated to the cloud database.
If though desirable I could make a simpler PR which just fixed the typo (currently if curves2 does not exist it uses curves, but assignes it to ecbd instead of ecdb -- and no-one noticed before now).

**Comment by davidfarmer** [comment]
Yes, we need to have someone listed for everything.
You can put me for the Picard Maass forms.

**Comment by roed314** [comment]
Yep, I'm happy to be the contact, and I should have some time this week to
push the work from the May workshop.

**Comment by AndrewVSutherland** [comment]
PR #1556 consolidates the siegel_modular_forms_experimental database (which contained only one collection) into the siegel_modular_forms database.  Once it has been pushed to beta and prod we can delete the siegel_modular_forms database from both Warwick and the cloud.

**Comment by JohnCremona** [comment]
PR#1556 has now been merged.

**Comment by AndrewVSutherland** [comment]
I have updated the documentation in https://github.com/LMFDB/lmfdb-inventory for all the databases listed in https://github.com/LMFDB/lmfdb/issues/1431#issuecomment-225474774 that are not tagged for deletion, with the exception of upload and userdb (which do not appear on http://www.lmfdb.org/knowledge/api but perhaps should still be documented).  You can find the table at https://github.com/LMFDB/lmfdb-inventory/blob/master/README.md (or just scroll down from the LMFDB/lmfdb-inventory home page).

With the exception of the MaassWaveForms database, which I have not had time to investigate but, like modularforms2, appears to contain many collections that are not strictly necessary (perhaps @sehlen or @fredstro can clarify and document them at https://github.com/LMFDB/lmfdb-inventory/blob/master/db-MaassWaveForms.md), I have gone through each database and listed all the collections that I believe are required by the LMFDB web server.  Any collection not listed is a candidate for deletion, at least from the cloud mongo db and possibly also from the Warwick mongo db.

If you are the contact person for a database (or even if you are not), please review what I have done, make any appropriate corrections, and add any missing documentation.  If there are collections that you think should be listed but are not, please add them, both to the README.md file and to the db-database.md file for the database.

**Comment by AndrewVSutherland** [comment]
The current list of databases that are candidates for deletion are:
- Lfunction (see issue #https://github.com/LMFDB/lmfdb/issues/1456)
- ap_statistics
- siegel_modular_forms_experimental (after PR #1556 is pushed to prod)
- test
If there is anything in the ap_statistics or test databases that you think should not be deleted, please speak now! (of course we will have backups).

I will post a list of individual collections in other databases that are candidates for deletion (either from the cloud or from both the cloud and Warwick) later today.

**Comment by JohnCremona** [comment]
Thanks @AndrewVSutherland -- having that table there is incredibly useful.  In the first column some of the links take me directly to one of the inventory documents in editor mode -- is that intentional?

**Comment by AndrewVSutherland2** [comment]
No, I'll fix those.

**Comment by AndrewVSutherland** [comment]
@JohnCremona Fixed.

**Comment by edgarcosta** [comment]
@AndrewVSutherland this is great!

Should we move your comment to somewhere else more visible? 
Perhaps, the maillist?
at the moment it is buried under other 80+ comments, and I'm afraid most people aren't following this thread.

**Comment by AndrewVSutherland** [comment]
@edgarcosta That probably makes sense, but let's wait until I have had a chance to look through the MaassWaveForms database and get a candidate list of collections for it as well.

**Comment by AndrewVSutherland** [comment]
The database siegel_modular_forms_experimental has been removed from both warwick and the cloud (the one collection samples it contained is now siegel_modular_forms.experimental_samples).

**Comment by AndrewVSutherland** [comment]
I added a new /stats option to the api interface that generates a table of mongo db collections ordered by size and/or number of objects; you can check it out at:

http://www.lmfdb.org/api/stats
http://beta.lmfdb.org/api/stats

One thing that jumps out is that on beta the single largest collection (about 300 GB) is ap.chunks in modularforms2 and the third largest is vector_on_basis.chunks (at 135 GB).  Neither of these collections is present on www.lmfdb.org.  Is there any reason to keep them on beta?

**Comment by AndrewVSutherland** [comment]
Following the discussion on #1294 and a review of the outstanding issues on LMFD/invenetory, I propose that we delete the following collections from the mongo db on lmfdb.warwick.ac.uk and then close this issue:

http://beta.lmfdb.org/api/modularforms2/ap.chunks/
http://beta.lmfdb.org/api/modularforms2/ap.files/
http://beta.lmfdb.org/api/modularforms2/Atkin_Lehner.chunks/
http://beta.lmfdb.org/api/modularforms2/Atkin_Lehner.files/
http://beta.lmfdb.org/api/modularforms2/vector_on_basis.chunks/
http://beta.lmfdb.org/api/modularforms2/vector_on_basis.files/
http://beta.lmfdb.org/api/modularforms2/Newform_factors.chunks/
http://beta.lmfdb.org/api/modularforms2/Newform_factors.files/
http://beta.lmfdb.org/api/modularforms2/Modular_symbols.chunks/
http://beta.lmfdb.org/api/modularforms2/Modular_symbols.files/
http://beta.lmfdb.org/api/modularforms2/WebNewforms.chunks/
http://beta.lmfdb.org/api/modularforms2/WebNewforms.files/
http://beta.lmfdb.org/api/modularforms2/webmodformspace_dimension/
http://beta.lmfdb.org/api/modularforms2/maeda/
http://beta.lmfdb.org/api/modularforms2/converted_E/
http://beta.lmfdb.org/api/ap_statistics/ap_data/

None of these collections are present on www.lmfdb.org, and none of them are referenced anywhere in the LMFDB code base;  @sehlen, @fredstro does this list look OK to you?  If there are any collections listed above that you think should be kept, please say so now (they will still be available on backups).

We also plan to eventually remove the database Lfunction (as opposed to Lfunctions), but this is covered under a separate issue #1456 and will be done only after references to it in the code are removed.

Once this is done we can proceed with #1294 and close that issue as well.

**Comment by fredstro** [comment]
As I said around May when this question came up in another thread or email discussion about modular forms (they are all so full of information now that I can't find which one it was )
As far as I know you xan delete all of them. I actually thought they were deleted after I said this the last time but didn't bother to check.

**Comment by sehlen** [comment]
Fredrik, this is about Warwick now. I don't have time to look at it right
now but we don't want the modular symbols collections etc to be deleted
there without moving/copying them to another mongodb first, I guess...

**Comment by AndrewVSutherland** [comment]
@fredstro Thanks for the confirmation!

**Comment by AndrewVSutherland** [comment]
@sehlen, @fredstro If you still need Modular_symbols.chunks and Modular_symbols.files, that's fine, I can remove them from the list of collections to be removed (they aren't all that big), I just need to know.

**Comment by AndrewVSutherland** [comment]
Given that we have not heard anything further from @sehlen or @fredstro I would propose that we go ahead and delete all the collections listed in https://github.com/LMFDB/lmfdb/issues/1431#issuecomment-233130105 except for modularforms2/ModularSymbols.* and plan to proceed with #1294 (perhaps this weekend @edgarcosta, @JohnCremona ?).

**Comment by JohnCremona** [comment]
Sounds good to me.  Perhaps do it to finish before the weekly backup starts at midnight Saturday GMT (which is 2300 BST so I guess 1800 Saturday on the East coast).

**Comment by edgarcosta** [comment]
FYI, I will be flying back to the US over this weekend.

**Comment by fredstro** [comment]
Ok, when I wrote my answer I didn't see (I was reading on my phone) that this was concerning the mongodb on lmfdb in Warwick since I didn't see any obvious reason for deleting any of the mentioned collections from that server.
After the previous discussions in May (around issue #1376) I believed that we had sufficient space to store these collections there and that the conclusion was that we decided on which collections are needed for the website and these were copied to the cloud while the lmfdb.warwick server could continue to contain collections with "primitive" research data not necessarily presented on the website. 
All of the mentioned collections except the last 3 (maeda, ap_data and converted_E) are used to construct the WebNewForms but they are not accessed from the website so if the new url e is that lmfdb.warwick.ac.uk should only contain data for the website (exactly as the cloud servers) then they can safely be removed. 
Since it seems that you are running out of space  we  (Stephan and I) can try to find somewhere else to store the modular forms data.  Since my last personal backup was in May I am now making another dump so we don't lose anything computed by Stephan after that point.

After Stephan's latest rewriting I wasn't sure exactly which collections contains what and which ones we would like to keep. I had hoped we had time for Stephan to finish his move to Germany and get settled before the storage issues became critical. 

There are too many messages from github at the moment so my gmail just dumps everything into spam even though I try to mark them as non-spam. It would be good if there was a setting so that some  of these messages only got  sent to the release managers so I could actually follow the relevant discussions (as I don't want to sync my massive spam folder to my phone).

**Comment by AndrewVSutherland** [comment]
@fredstro I will delete the last 3, but are you sure you really need all the others?

For example, are the collections

webnewforms.chunks and webnewforms2.chunks and WebNewForms.chunks
webnewforms.files and webnewforms2.files and WebNewForms.files
webmodformspace.chunks and webmodformspace2.chunks and WebModformspace.chunks
webmodformspace.files and webmodformspace2.files and WebModformspace.files

all necessary?  Perhaps just 2 (rather than 3) variants of each of these would suffice?

BTW, you can browse the full list of collections on the Warwick mongo db modularforms2 database at http://beta.lmfdb.org/api/ (scroll down to see modularforms2).

**Comment by AndrewVSutherland** [comment]
@fredstro One thing you can do to reduce the number of github e-mails you get is to go to https://github.com/settings/notifications and turn off e-mails for repositories you are watching (you will still get e-mails for conversations you are participating in (like this one), and for comments that @mention you explicitly).  You can also unclick the notifications for Pull Request Reviews and Pull Request Pushes (I think you will still get notifications regarding pull requests you initiate).

**Comment by JohnCremona** [comment]
It is not that we are running out of space on the Warwick server, but we still want to delete things which are not needed, since not only do they use up space but so do their copies on the weekly backups.  We do have more disks which can be brought online but I am delaying that until a lot of other things have finished since last time I did that it was quite disruptive.

**Comment by fredstro** [comment]
Ok. 
I was responding to the collections you (Drew) listed above which did not include the lower case versions.
@JohnCremona I see, I can understand that although the largest collection is that of aps which is quite important (since there might be a lot of them which are not converted to the web format yet) . 
Most of the 'random' collections are just used for testing and could probably be removed but I wouldn't do it before making sure that Stephan doesn't need them.

 I can only tell you which collections were used before Stephan's May refactoring (I am not 100% certain he did not change the names or introduced anything new). Namely the following:

* Hecke eigenvalues for newforms:
http://beta.lmfdb.org/api/modularforms2/ap.chunks/
http://beta.lmfdb.org/api/modularforms2/ap.files/

* Atkin lehner eigenvalues:
http://beta.lmfdb.org/api/modularforms2/Atkin_Lehner.chunks/
http://beta.lmfdb.org/api/modularforms2/Atkin_Lehner.files/

* A vector which describes the newform on a given basis:
http://beta.lmfdb.org/api/modularforms2/vector_on_basis.chunks/
http://beta.lmfdb.org/api/modularforms2/vector_on_basis.files/

* Basis matrix and some other data for the newform factors:
http://beta.lmfdb.org/api/modularforms2/Newform_factors.chunks/
http://beta.lmfdb.org/api/modularforms2/Newform_factors.files/

* Basis matrix and other data for the ambient space of modular symbols
http://beta.lmfdb.org/api/modularforms2/Modular_symbols.chunks/
http://beta.lmfdb.org/api/modularforms2/Modular_symbols.files/

* I don't think this is used but Stephan will know:
http://beta.lmfdb.org/api/modularforms2/WebNewforms.chunks/
http://beta.lmfdb.org/api/modularforms2/WebNewforms.files/

* Table of dimensions. This is actually referenced in LMFDB in lmfdb/modular_forms/elliptic_modular_forms/views/emf_render_web_modform_space_gamma1.py: 69 (but the variable dim_table is actually not used later in the function so this can be probably be removed from the function and consequently also the collection from the database)
http://beta.lmfdb.org/api/modularforms2/webmodformspace_dimension/

I don't recognise a lot of the collections in /api/ so I wouldn't remove any of them without checking that Stephan is done with the testing/checking/recomputations which began in May. 
@sehlen : from my point of view, I don't need to keep any of the collections other than the ones mentioned above together with the ones used for the webobjects (whatever names they may have)

**Comment by JohnCremona** [comment]
@fredstro It would be great to put some of the information above into https://github.com/LMFDB/lmfdb-inventory/blob/master/db-modularforms2.md -- even a short description without the detail (which may not even make sense for thse files/chunks things).

**Comment by AndrewVSutherland** [comment]
@fredstro Thank your for the clarification this is very helpful!

Among the larger collections (those > 1GB, there are also many smaller ones that I expect can be deleted but there is no rush to deal with those), the only remaining candidates for deletion are:

http://beta.lmfdb.org/api/modularforms2/WebNewforms.chunks/
http://beta.lmfdb.org/api/modularforms2/WebNewforms.files/

which look to me like old versions of webnewforms.* (and there is also webnewforms2.*).

@sehlen Can you confirm that it is OK to delete these two collections (or alternatively, confirm that they are needed)?  Once that is done I think we can close this issue.

**Comment by fredstro** [comment]
Ok, I will try to fill in some of the documentation there. I thought it was just for the ones which are used on the website (which I can write what I think it is but it might not be correct since Stephan might have changed something).
I also thought it only applied for collections which could be obtained in json format I didn't think about all the gridfs collections. 
I am not sure what we really intend with the api but if we really want users to interface with it we should probably give an openapi/swagger spec of all endpoints and only allow access to fully described endpoints.

**Comment by JohnCremona** [comment]
Thanks -- obviously this is most important to have the details for collections in use on the website, but for anything which is not strictly temporary (say, you expect it to still be useful in a few months' time) then having a small note about it would stop others from asking you if it can be deleted!

**Comment by sehlen** [comment]
> On Jul 20, 2016, at 09:50, Andrew Sutherland <notifications@github.com> wrote:
> 
> @fredstro <https://github.com/fredstro> Thank your for the clarification this is very helpful!
> 
> Among the larger collections (those > 1GB, there are also many smaller ones that I expect can be deleted but there is no rush to deal with those), the only remaining candidates for deletion are:
> 
> http://beta.lmfdb.org/api/modularforms2/WebNewforms.chunks/ <http://beta.lmfdb.org/api/modularforms2/WebNewforms.chunks/>
> http://beta.lmfdb.org/api/modularforms2/WebNewforms.files/ <http://beta.lmfdb.org/api/modularforms2/WebNewforms.files/>
> 

I don’t think they are used. In fact, I did not add new collections (I think).
Maybe @fredstro could take a quick look at the data. If you have a backup, Fredrik, then that’s also fine, I guess.

> which look to me like old versions of webnewforms.* (and there is also webnewforms2.*).
> 
I think @fredstro used webnewforms2 when he started cleaning up data last year before the release.

> @sehlen <https://github.com/sehlen> Can you confirm that it is OK to delete these two collections (or alternatively, confirm that they are needed)? Once that is done I think we can close this issue.
> 
> 

I think so. But, as I said, I cannot really start looking at these things now as there are many things to finish before my family and I move back to Germany on Saturday.
Also, @fredstro, if you start documenting, please have a look at my fork of the inventory repo, where I started the documentation to the best of my knowledge.

**Comment by JohnCremona** [comment]
Indeed https://github.com/sehlen/lmfdb-inventory says that it is 5 commits ahead of the master, and all those commits are called "Update db-modularforms2.md".  Pulling them into the master would be a good place to start.

**Comment by fredstro** [comment]
ah, damn it, I started with the same collections as you did but with the format similar to the L-functions.

**Comment by sehlen** [comment]
Sorry, but I mentioned this at least three times now.

**Comment by fredstro** [comment]
That's alright, I just thought that anything you added would have been pushed to master.

**Comment by sehlen** [comment]
> On Jul 20, 2016, at 12:14, Fredrik Strömberg <notifications@github.com> wrote:
> 
> That's alright, I just thought that anything you added would have been pushed to master.
> 
No, I didn’t issue a pull request because I was not nearly done yet and the information is partly “wrong” because of that.

**Comment by fredstro** [comment]
After checking I dropped the following collections: 
* test
* dimension_table_old
* webnewforms2 + .files + .chunks
* webmodformspace2 + .files + .chunks
* WebModformspace.files + .chunks
* WebNewforms.files + .chunks
* file_checked
* ambient_mongo_checked'
* checked_stuff
* webmodformspace_errors
* character_orbits.files + .chunks 
* aps_mongo_checked
* dimensions (this was used by the now obsolete (I think) class DimensionTable in emf_classes.py
the remaining collections are used (for the moment) and should be kept and they are all documented (if I missed some let me know)

**Comment by AndrewVSutherland** [comment]
@fredstro Thank you!  I am now happy to close this issue.
